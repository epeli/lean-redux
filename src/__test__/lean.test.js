import React from "react";
import renderer from "react-test-renderer";
import {createStore} from "redux";
import {Provider} from "react-redux";

import {get} from "lodash/fp";

import leanReducer, {connectLean} from "../lean";



test("basic updating", () => {

    const store = createStore(leanReducer);
    var update = null;
    const My = ({name, changeName}) => {
        update = changeName;
        return <div>Hello {name}</div>;
    };

    const Connected = connectLean({
        scope: "ascope",
        defaults: {
            name: "esa",
        },
        updates: {
            changeName() {
                return {name: "matti"};
            },
        },
    })(My);

    const Main = () => {
        return (
            <Provider store={store}>
                <Connected />
            </Provider>
        );
    };

    expect(get(["ascope", "name"], store.getState())).toBe(undefined);

    const component = renderer.create(<Main />);
    expect(component.toJSON()).toMatchSnapshot();
    update();
    expect(component.toJSON()).toMatchSnapshot();

    expect(get(["ascope", "name"], store.getState())).toBe("matti");

});

test("updating with map state", () => {

    const store = createStore(leanReducer);
    var update = null;
    const My = ({name, changeName}) => {
        update = changeName;
        return <div>Hello {name}</div>;
    };

    const Connected = connectLean({
        scope: "ascope",
        defaults: {
            name: "esa",
        },
        mapState(state) {
            return {name: state.name.toUpperCase()};
        },
        updates: {
            changeName() {
                return {name: "matti"};
            },
        },
    })(My);

    const Main = () => {
        return (
            <Provider store={store}>
                <Connected />
            </Provider>
        );
    };

    expect(get(["ascope", "name"], store.getState())).toBe(undefined);

    const component = renderer.create(<Main />);
    expect(component.toJSON()).toMatchSnapshot();
    update();
    expect(component.toJSON()).toMatchSnapshot();

    expect(get(["ascope", "name"], store.getState())).toBe("matti");

});

test("can use function to update the state", () => {

    const store = createStore(leanReducer);
    var update = null;
    const Counter = ({count, inc}) => {
        update = inc;
        return <div>{count}</div>;
    };

    const Connected = connectLean({
        scope: "ascope",
        defaults: {
            count: 0,
        },
        updates: {
            inc() {
                return {count: i => i+1};
            },
        },
    })(Counter);

    const Main = () => {
        return (
            <Provider store={store}>
                <Connected />
            </Provider>
        );
    };

    expect(get(["ascope", "count"], store.getState())).toBe(undefined);

    const component = renderer.create(<Main />);
    expect(component.toJSON()).toMatchSnapshot();
    update();
    expect(component.toJSON()).toMatchSnapshot();

    expect(get(["ascope", "count"], store.getState())).toBe(1);

});

test("pass scope as a prop", () => {

    const store = createStore(leanReducer);
    var update = null;

    const My = ({name, changeName}) => {
        update = changeName;
        return <div>Hello {name}</div>;
    };

    const Connected = connectLean({
        defaults: {
            name: "esa",
        },
        updates: {
            changeName() {
                return {name: "matti"};
            },
        },
    })(My);

    const Main = () => {
        return (
            <Provider store={store}>
                <Connected scope="propScope" />
            </Provider>
        );
    };

    expect(get(["propScope", "name"], store.getState())).toBe(undefined);

    const component = renderer.create(<Main />);
    expect(component.toJSON()).toMatchSnapshot();
    update();
    expect(component.toJSON()).toMatchSnapshot();

    expect(get(["propScope", "name"], store.getState())).toBe("matti");

});


test("do not recreate update function on parent prop changes", () => {
    const store = createStore(leanReducer);
    var update = null;

    const My = ({name, greeting, changeName}) => {
        update = changeName;
        return <div>{greeting} {name}</div>;
    };

    const Connected = connectLean({
        defaults: {
            name: "esa",
        },
        updates: {
            changeName() {
                return {name: "matti"};
            },
        },
    })(My);

    var setState = null;
    var Main = React.createClass({
        getInitialState() {
            return {greeting: "Hi"};
        },
        changeGreeting() {
        },
        render() {
            setState = this.setState.bind(this);
            return (
                <Provider store={store}>
                    <Connected scope="propScope" greeting={this.state.greeting} />
                </Provider>
            );
        },
    });

    const component = renderer.create(<Main />);
    expect(component.toJSON()).toMatchSnapshot();
    var prevUpdate = update;

    setState({greeting: "Hellou"});
    expect(component.toJSON()).toMatchSnapshot();

    expect(update).toBe(prevUpdate);

});

