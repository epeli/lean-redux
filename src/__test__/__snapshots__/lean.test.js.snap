exports[`test all props are passed to the update functions 1`] = `
<div>
  Hello 
  esa
</div>
`;

exports[`test all props are passed to the update functions 2`] = `
<div>
  Hello 
  suuronen
</div>
`;

exports[`test basic updating 1`] = `
<div>
  Hello 
  esa
</div>
`;

exports[`test basic updating 2`] = `
<div>
  Hello 
  matti
</div>
`;

exports[`test can use function to update the state 1`] = `
<div>
  0
</div>
`;

exports[`test can use function to update the state 2`] = `
<div>
  1
</div>
`;

exports[`test do not recreate update function on parent prop changes 1`] = `
<div>
  Hi
   
  esa
</div>
`;

exports[`test do not recreate update function on parent prop changes 2`] = `
<div>
  Hellou
   
  esa
</div>
`;

exports[`test lodash strings paths are not expanded in mapState 1`] = `
<div>
  hello 
  underscore
</div>
`;

exports[`test lodash strings paths are not expanded in mapState 2`] = `
<div>
  hello 
  lodash
</div>
`;

exports[`test lodash strings paths are not expanded in mapState 3`] = `
Object {
  "foo.bar": Object {
    "name": "lodash",
  },
}
`;

exports[`test pass scope as a prop 1`] = `
<div>
  Hello 
  esa
</div>
`;

exports[`test pass scope as a prop 2`] = `
<div>
  Hello 
  matti
</div>
`;

exports[`test prop scope change generates new update function 1`] = `
<div>
  Hello 
  matti
</div>
`;

exports[`test prop scope change generates new update function 2`] = `
<div>
  Hello 
  matti
</div>
`;

exports[`test updating with map state 1`] = `
<div>
  Hello 
  ESA
</div>
`;

exports[`test updating with map state 2`] = `
<div>
  Hello 
  MATTI
</div>
`;
